### PR TITLE
RKE host structure and config changes

### PR DESCRIPTION
**Pull request by galal-hussein**
https://github.com/rancher/rke/issues/53
https://github.com/rancher/rke/issues/52
https://github.com/rancher/rke/issues/50
https://github.com/rancher/rke/issues/41
https://github.com/rancher/rke/issues/36

- `IP` to `Address`
- `AdvertiseAddress` to `InternalAddress`
- `AdvertisedHostname` to `HostnameOverride`

Hostname will be optional, and if not set it will be equal to Address.

### CODE REVIEW

**Comment by galal-hussein** [comment]
@alena1108 @moelsayed Can you please review?

File: hosts/hosts.go [hunk 1]
```diff
@@ -90,7 +90,7 @@ func IsHostListChanged(currentHosts, configHosts []Host) bool {
 	for _, host := range currentHosts {
 		found := false
 		for _, configHost := range configHosts {
-			if host.AdvertisedHostname == configHost.AdvertisedHostname {
+			if host.Address == configHost.Address {
 				found = true
```
Comment (alena1108) [inline hosts/hosts.go]: we should break when found=true

File: cluster/validation.go [hunk 2]
```diff
@@ -13,16 +15,79 @@ func (c *Cluster) ValidateCluster() error {
...
+			if role != services.ETCDRole && role != services.ControlRole && role != services.WorkerRole {
+				return fmt.Errorf("Role [%s] for host (%d) is not recognized", role, i+1)
+			}
+		}
+		bothRoles := make(map[string]bool)
+		for _, role := range host.Role {
+			if role == services.ControlRole {
+				bothRoles[services.ControlRole] = true
+			}
+			if role == services.WorkerRole {
```
Comment (alena1108) [inline cluster/validation.go]: should be elseif

File: cluster/validation.go [hunk 3]
```diff
@@ -13,16 +15,79 @@ func (c *Cluster) ValidateCluster() error {
...
+		}
+		if len(host.Role) == 0 {
+			return fmt.Errorf("Role for host (%d) is not provided", i+1)
+		}
+		for _, role := range host.Role {
+			if role != services.ETCDRole && role != services.ControlRole && role != services.WorkerRole {
+				return fmt.Errorf("Role [%s] for host (%d) is not recognized", role, i+1)
+			}
+		}
+		bothRoles := make(map[string]bool)
```
Comment (alena1108) [inline cluster/validation.go]: any reason why we use a map here? there are only 2 values, workerRole and controlRole, may be can use bool instead to simplify the outlook?

File: cluster/validation.go [hunk 4]
```diff
@@ -13,16 +15,79 @@ func (c *Cluster) ValidateCluster() error {
 	if len(c.EtcdHosts) == 0 {
```
Comment (alena1108) [inline cluster/validation.go]: Should we validate number of etcd hosts (it should be an odd number)?

File: cluster/validation.go [hunk 5]
```diff
@@ -13,16 +15,79 @@ func (c *Cluster) ValidateCluster() error {
...
-func validateServicesOption(c *Cluster) error {
+func validateAuthOptions(c *Cluster) error {
+	if c.Authentication.Strategy != DefaultAuthStrategy {
+		return fmt.Errorf("Authentication strategy [%s] is not supported", c.Authentication.Strategy)
+	}
+	return nil
+}
+
+func validateNetworkOptions(c *Cluster) error {
+	if c.Network.Plugin != "flannel" && c.Network.Plugin != "calico" && c.Network.Plugin != "canal" {
```
Comment (alena1108) [inline cluster/validation.go]: I suggest you define the constants for the network plugins.

File: hosts/hosts.go (hunk 1 above)
Comment (galal-hussein) [reply hosts/hosts.go]: fixed it

File: cluster/validation.go [hunk 6]
```diff
@@ -3,26 +3,88 @@ package cluster
...
 )
 
 func (c *Cluster) ValidateCluster() error {
 	// make sure cluster has at least one controlplane/etcd host
 	if len(c.ControlPlaneHosts) == 0 {
 		return fmt.Errorf("Cluster must have at least one control plane host")
 	}
-	if len(c.EtcdHosts) == 0 {
-		return fmt.Errorf("Cluster must have at least one etcd host")
+	if len(c.EtcdHosts)%2 == 0 {
```
Comment (alena1108) [inline cluster/validation.go]: what if len is 0?

**Pull request merged**